### DISCUSSION TITLE
Type conversion problem

### DISCUSSION THREAD

**Discussion by feketeimre** [pythonnet/pythonnet#2375] [General]
I am trying to interface with a .NET dll to read stuff from a camera and i have trouble figuring out the types for a function call. Admittedly i have just started using pythonnet so my problem might be trivial.

Here is the truncated version of my program:
```python
import clr
from System import *
clr.AddReference('PylonC.NET')
from PylonC.NET import Pylon
from PylonC.NET import PylonBuffer
from PylonC.NET import PylonGrabResult_t

hDev = Pylon.CreateDeviceByIndex(0)
imgBuf = PylonBuffer[Byte](None) #PylonBuffer<Byte> imgBuf = null; Is None correct here ?
grabResult = PylonGrabResult_t()

ret = Pylon.DeviceGrabSingleFrame(hDev, Int32(0), imgBuf, grabResult, UInt32(500))
# the above raises an exception:
# System.ArgumentException: PylonC.NET.PylonBuffer`1[System.Byte] value cannot be converted to PylonC.NET.PylonBuffer`1[T] in method
# Boolean DeviceGrabSingleFrame[T](PylonC.NET.PYLON_DEVICE_HANDLE, Int32, PylonC.NET.PylonBuffer`1[T] ByRef, PylonC.NET.PylonGrabResult_t ByRef, UInt32)
```
I have doubts if the imgBuf = PylonBuffer[Byte](None) is equivalent to the original c: PylonBuffer<Byte> imgBuf = null;
But mainly the type conversion error is what i have no idea why it doesnt work. I am certain that Byte is a valid type for PylonBuffer.
With dnSpy i looked into the problematic function and found this:
![grab 2024-05-07 210235](https://github.com/pythonnet/pythonnet/assets/15861028/ce647da8-96ed-48d2-af65-768e2762d6e6)

Also im not sure how to deal with the ref and out keywords. 

Any help is welcomed,
Thank you

**Comment by yezhou2000** [comment]
Hi I got the same question. Did you figure it out? Like how to convert imgBuf into the ref type needed in C# function? Thank you!

**Reply by feketeimre** [reply]
Hi, unfortunately no, i stopped trying.

**Comment by feketeimre** [comment]
I spent time on this and managed to figure it out, sort of...
The function expects a generic type [T] but i have [Byte]. When passing Byte to this function it gives the exception as previously mentioned:
```python
# System.ArgumentException: PylonC.NET.PylonBuffer`1[System.Byte] value cannot be converted to PylonC.NET.PylonBuffer`1[T] in method .....
```

Apparently, you can construct a specific type of the function like:
```python
grabFrame = Pylon.DeviceGrabSingleFrame[Byte]
ret = grabFrame(hDev, Int32(0), imgBuf, grabResult, UInt32(500))
```
In most cases i think this would work, but in my case i need to pass imgBuf and grabResult by reference that cannot be done in pythonnet.
